### PR TITLE
Avoid connect repo when individual file is open.

### PR DESCRIPTION
**Pull request by syedgulahmed**
**Testing Instructions**

- Open a project or directory that is not synced, you will notice popup will appear to sync the project.
- Open any individual file and before you open an individual file make sure you close already open project from File menu using 'Close Project'. When you will open an individual file you will notice sync popup is not being shown.

### CODE REVIEW

File: src/main/java/org/intellij/sdk/codesync/Utils.java [hunk 1]
```diff
@@ -45,6 +45,16 @@ public static Boolean shouldSkipEvent(String repoPath) {
         }
     }
 
+    public static boolean isIndividualFileOpen(Project project) {
+
+        VirtualFile[] virtualFiles = ProjectUtils.getAllContentRoots(project);
+        if (virtualFiles[0].getPath().equals(project.getBasePath())) {
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/Utils.java:51]: What if we have something like 
```
mydir/
    mydir (A file but having same name as parent directory)
    anotherFile.txt
```

Comment (Basit-qc) [reply src/main/java/org/intellij/sdk/codesync/Utils.java:51]: Will be great to add some unit tests for this function with possible edge cases

File: src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java [hunk 2]
```diff
@@ -149,7 +148,7 @@ public static void setupCodeSyncRepo(Project project, String repoPath, String re
             ConfigFile configFile = new ConfigFile(CONFIG_PATH);
 
             if (!configFile.isRepoActive(repoPath) || isSyncingBranch) {
-                String branchName = Utils.GetGitBranch(repoPath);
+                String branchName = Utils.GetGitBranch(repoPath); //TODO handle gracefully if only file open.
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java:151]: Probably no need to check branch if single file is opened.
```
...
if (!Utils.isIndividualFileOpen(project)) return;
String branchName = Utils.GetGitBranch(repoPath);
...
```

**Review by Basit-qc — CHANGES_REQUESTED** [review]

File: src/main/java/org/intellij/sdk/codesync/Utils.java (hunk 1 above)
Comment (syedgulahmed) [reply src/main/java/org/intellij/sdk/codesync/Utils.java:51]: > What if we have something like
> 
> ```
> mydir/
>     mydir (A file but having same name as parent directory)
>     anotherFile.txt
> ```

It will still work because if we are opening a single file as project on Intellij, it creating its base directory in temp directory of system. Otherwise it is same as root directory of folder.

File: src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java (hunk 2 above)
Comment (syedgulahmed) [reply src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java:151]: Updating it.

File: src/main/java/org/intellij/sdk/codesync/Utils.java (hunk 1 above)
Comment (Basit-qc) [reply src/main/java/org/intellij/sdk/codesync/Utils.java:51]: I understand that but the example I shared is different and seems like your logic will fail on it. In my example I have a directory which has 2 files, 1 has same name as directory, other file has different name. In this case 
```
virtualFiles[0].getPath().equals(project.getBasePath())
```
will return true, right? Since you are only comparing 1st file path with directory path. 
I think we need additional check as 
```
virtualFiles.length == 1 && virtualFiles[0].getPath().equals(project.getBasePath())
```
Thoguhts?

Comment (syedgulahmed) [reply src/main/java/org/intellij/sdk/codesync/Utils.java:51]: No, I'm not comparing file path! The method I'm using returns content root, which is basically root directory for the folder. I will update the code and make it clear.

Comment (Basit-qc) [reply src/main/java/org/intellij/sdk/codesync/Utils.java:51]: I am talking about multi-modules, for that I think we need to check length of contentRoots, if there are multiple modules open, it should return false. 
Unit tests will defineltiy catch these issues before hand. Kindly add a few unit tests for different cases, thanks

Comment (Basit-qc) [reply src/main/java/org/intellij/sdk/codesync/Utils.java:51]: Thoughts? @saleem-latif

File: src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java [hunk 3]
```diff
@@ -145,6 +144,12 @@ public void run(@NotNull ProgressIndicator progressIndicator) {
     }
 
     public static void setupCodeSyncRepo(Project project, String repoPath, String repoName, CodeSyncProgressIndicator codeSyncProgressIndicator, boolean skipSyncPrompt, boolean isSyncingBranch) {
+
+        if(Utils.isIndividualFileOpen(project)){
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java:148]: When will code reach here? I mean "Connect" option is disabled for singe-file-opened-as-project, right?

File: src/main/java/org/intellij/sdk/codesync/Utils.java [hunk 4]
```diff
@@ -45,6 +45,18 @@ public static Boolean shouldSkipEvent(String repoPath) {
         }
     }
 
+    public static boolean isIndividualFileOpen(Project project) {
+
+        VirtualFile[] virtualFiles = ProjectUtils.getAllContentRoots(project);
+
+        if (virtualFiles.length > 1) {
+            return false;
+        } else {
```
Comment (Basit-qc) [inline src/main/java/org/intellij/sdk/codesync/Utils.java:54]: else block is not required here

File: CHANGELOG.md [hunk 5]
```diff
@@ -4,6 +4,9 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [3.23.5] - 2023-08-22
```
Comment (Basit-qc) [inline CHANGELOG.md:7]: Need to correct the date

File: src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java (hunk 3 above)
Comment (syedgulahmed) [reply src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java:148]: This method is being called on startup and it will also produce error message to user when they try to click "Connect" button.

Comment (Basit-qc) [reply src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java:148]: Point is, why would user even see the "Connect" button? It should not be visible, right?

File: src/main/java/org/intellij/sdk/codesync/Utils.java [hunk 6]
```diff
@@ -45,6 +45,18 @@ public static Boolean shouldSkipEvent(String repoPath) {
         }
     }
 
+    public static boolean isIndividualFileOpen(Project project) {
+
+        VirtualFile[] virtualFiles = ProjectUtils.getAllContentRoots(project);
```
[the remark this answers is not in the document]
Comment (saleem-latif) [reply src/main/java/org/intellij/sdk/codesync/Utils.java:50]: You will have to use something like 
```
String[] filePaths = FileUtils.listFiles(repoPath);
```
to get the list of all files that are opened. but this is a time expensive process, so I would suggest you find an alternative approach that uses Intellij's VirtualFileSystem to make it faster. You can look into  https://plugins.jetbrains.com/docs/intellij/file-based-indexes.html#word-index and https://github.com/JetBrains/intellij-community/blob/idea/232.9559.62/platform/indexing-api/src/com/intellij/psi/search/FilenameIndex.java#L35  for better approaches.

File: src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java (hunk 3 above)
Comment (saleem-latif) [reply src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java:148]: @syedgulahmed We can disable the connect button from `CodeSyncSetupAction` here https://github.com/codesyncapp/codesync-intellij/blob/master/src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetupAction.java#L57 is an example.

File: src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java [hunk 7]
```diff
@@ -168,7 +173,7 @@ public static void setupCodeSyncRepo(Project project, String repoPath, String re
                 // That is why this boolean is placed here.
                 boolean shouldSyncRepo = false;
                 // Do not ask user to sync repo, if it is already in progress.
-                if (!reposBeingSynced.contains(repoPath)) {
+                if (!reposBeingSynced.contains(repoPath) && !Utils.isIndividualFileOpen(project)) {
```
Comment (saleem-latif) [inline src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java:176]: I do not think we need this here, it will cause more bugs in this place.

File: src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetupAction.java [hunk 8]
```diff
@@ -69,9 +70,11 @@ public void update(@NotNull AnActionEvent e) {
                 presentation.setDescription("Connect repo with codeSync");
             }
 
-            String repoPath = ProjectUtils.getRepoPath(project);
-            PluginState pluginState = StateUtils.getState(repoPath);
-            e.getPresentation().setEnabled(!pluginState.syncInProcess);
+            if(!Utils.isIndividualFileOpen(project)){
```
Comment (saleem-latif) [inline src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetupAction.java:73]: Why do we need it here? I do not think we should add it here.

**Review by saleem-latif — CHANGES_REQUESTED** [review]
@syedgulahmed We should add the checks at the entry points only, like https://github.com/codesyncapp/codesync-intellij/blob/master/src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetupAction.java#L28 and https://github.com/codesyncapp/codesync-intellij/blob/master/src/main/java/org/intellij/sdk/codesync/PopulateBuffer.java#L222
I think these are the only 2 places that will need the check. You will not need to change anything else.

**Review by Basit-qc — APPROVED** [review]

File: src/main/java/org/intellij/sdk/codesync/Utils.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Basit-qc) [reply src/main/java/org/intellij/sdk/codesync/Utils.java:50]: Thanks for the insights @saleem-latif. Wondering how was that was working as a fix? @syedgulahmed How did you test that?

File: src/test/java/org/intellij/sdk/codesync/UtilsTest.kt [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.VirtualFile
+import org.intellij.sdk.codesync.utils.ProjectUtils
+import org.junit.jupiter.api.Test
+import org.mockito.Mockito.*
+import kotlin.test.assertFalse
+import kotlin.test.assertTrue
+
+class UtilsTest {
```
Comment (Basit-qc) [inline src/test/java/org/intellij/sdk/codesync/UtilsTest.kt:11]: If getAllContentRoots isn't the correct fix, Why our unit tests catch the issue? @syedgulahmed

File: src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java (hunk 7 above)
Comment (syedgulahmed) [reply src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java:176]: Yes, removed it.

File: src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java (hunk 3 above)
Comment (syedgulahmed) [reply src/main/java/org/intellij/sdk/codesync/codeSyncSetup/CodeSyncSetup.java:148]: Well `setupCodeSyncRepo` is called in `setupCodeSyncRepoAsync` method that method has calls in in OpenCloseListener  and PopulateBuffer.

File: src/main/java/org/intellij/sdk/codesync/ProjectOpenCloseListener.java [hunk 10]
```diff
@@ -105,6 +106,12 @@ public void projectOpened(@NotNull Project project) {
       // Populate state for all the opened modules. Module is the term used for projects opened using "Attach" option
       // in the IDE open dialog box.
       for (VirtualFile contentRoot: contentRoots) {
+
+        if(Utils.isIndividualFileOpen(contentRoot.getPath())){
+          NotificationManager.notifyError(INDIVIDUAL_FILE_OPENED);
```
Comment (saleem-latif) [inline src/main/java/org/intellij/sdk/codesync/ProjectOpenCloseListener.java:111]: You can skip the notification here. The reason is that this code runs in the background and does not execute in response to user input.

User might be confused as to why we are saying that repo can not be synced because it is a file since he did not ask to sync the repo.

File: src/main/java/org/intellij/sdk/codesync/Utils.java [hunk 11]
```diff
@@ -45,6 +45,11 @@ public static Boolean shouldSkipEvent(String repoPath) {
         }
     }
 
+    public static boolean isIndividualFileOpen(String contentRootPath) {
+        File file = new File(contentRootPath);
```
Comment (saleem-latif) [inline src/main/java/org/intellij/sdk/codesync/Utils.java:49]: Awesome 👍🏼 I think this is better than my suggestion. 
Make sure to test out all the flows though.

**Review by saleem-latif — APPROVED** [review]
Just one comment, once that is addressed it is good to go.

File: src/main/java/org/intellij/sdk/codesync/Utils.java (hunk 11 above)
Comment (syedgulahmed) [reply src/main/java/org/intellij/sdk/codesync/Utils.java:49]: Also, I believe we don't need to put check in detectBranchChange in populate anymore because it will not be reachable as single file repo will not be already synced.
